pluginsdk/support/bundlepublisherutil/bputil.go

```diff
@@ -0,0 +1,257 @@
+// Package bundlepublisherutil provides helper functions for plugins
+// implementing the BundlePublisher interface.
+// BundlePublisher plugins should use this package as a way to have a
+// standarized name for bundle formats in their configuration, and avoid the
+// re-implementation of bundle parsing logic of formats supported in this
+// package.
+package bundlepublisherutil
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/spiffe/go-spiffe/v2/bundle/spiffebundle"
+	"github.com/spiffe/go-spiffe/v2/spiffeid"
+	"github.com/spiffe/spire-plugin-sdk/proto/spire/plugin/types"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const (
+	BundleFormatUnset BundleFormat = iota
+	SPIFFE
+	PEM
+	JWKS
+)
+
+// Bundle represents a bundle that can be formatted in different formats.
+type Bundle struct {
+	bundle *types.Bundle
+
+	jwksBytes   []byte
+	pemBytes    []byte
+	spiffeBytes []byte
+}
+
+// KeyType represents the types of keys that are supported by the KeyManager.
+type BundleFormat int
+
+// BundleFormatFromString returns the BundleFormat corresponding to the provided
+// string.
+func BundleFormatFromString(s string) (BundleFormat, error) {
+	switch strings.ToLower(s) {
+	case "spiffe":
+		return SPIFFE, nil
+	case "jwks":
+		return JWKS, nil
+	case "pem":
+		return PEM, nil
+	default:
+		return BundleFormatUnset, fmt.Errorf("unknown bundle format: %q", s)
+	}
+}
+
+// NewBundle return a new *Bundle with the *types.Bundle provided.
+// Use the Bytes() function to get a slice of bytes with the bundle formatted in
+// the format specified.
+func NewBundle(pluginBundle *types.Bundle) *Bundle {
```

Comment:
It's hard to tell without knowing how this will be used in plugins, but do we imagine the Bytes() method being called more than once for a given bundle?

In either case, for callers who only need to format a single bundle one way, we may consider adding a top-level `FormatBundle(bundle *types.Bundle, format BundleFormat) ([]byte, error)` method. It can use the NewBundle/Bytes methods to do the formatting.

Reply:
Having a FormatBundle function makes sense.